packages/dds/tree/src/feature-libraries/default-schema/schemaChecker.ts

```diff
@@ -3,89 +3,74 @@
  * Licensed under the MIT License.
  */
 
-import { assert, unreachableCase } from "@fluidframework/core-utils/internal";
+import { unreachableCase } from "@fluidframework/core-utils/internal";
 import {
-	MapTree,
-	StoredSchemaCollection,
-	TreeFieldStoredSchema,
+	type MapTree,
+	type TreeFieldStoredSchema,
 	LeafNodeStoredSchema,
 	ObjectNodeStoredSchema,
 	MapNodeStoredSchema,
 	Multiplicity,
+	type SchemaAndPolicy,
+	getMapTreeField,
 } from "../../core/index.js";
-import { FullSchemaPolicy } from "../modular-schema/index.js";
 import { allowsValue } from "../valueUtilities.js";
+import { fail } from "../../util/index.js";
 
 export const enum SchemaValidationErrors {
 	NoError,
 	Field_KindNotInSchemaPolicy,
 	Field_IncorrectMultiplicity,
 	Field_NodeTypeNotAllowed,
-	LeafNode_MissingValue,
 	LeafNode_InvalidValue,
 	LeafNode_FieldsNotAllowed,
-	ObjectNode_FieldCountMismatch,
 	ObjectNode_FieldNotInSchema,
 	Node_MissingSchema,
 	UnknownError,
 }
 
 export function isNodeInSchema(
 	node: MapTree,
-	nodeSchemaCollection: StoredSchemaCollection,
-	schemaPolicy: FullSchemaPolicy,
+	schemaAndPolicy: SchemaAndPolicy,
 ): SchemaValidationErrors {
 	// Validate the schema declared by the node exists
-	const schema = nodeSchemaCollection.nodeSchema.get(node.type);
+	const schema = schemaAndPolicy.schema.nodeSchema.get(node.type);
 	if (schema === undefined) {
 		return SchemaValidationErrors.Node_MissingSchema;
 	}
 
 	// Validate the node is well formed according to its schema
 
 	if (schema instanceof LeafNodeStoredSchema) {
-		if (node.value === undefined) {
-			return SchemaValidationErrors.LeafNode_MissingValue;
-		}
 		if (node.fields.size !== 0) {
 			return SchemaValidationErrors.LeafNode_FieldsNotAllowed;
 		}
 		if (!allowsValue(schema.leafValue, node.value)) {
 			return SchemaValidationErrors.LeafNode_InvalidValue;
 		}
 	} else if (schema instanceof ObjectNodeStoredSchema) {
-		if (node.fields.size !== schema.objectNodeFields.size) {
-			return SchemaValidationErrors.ObjectNode_FieldCountMismatch;
-		}
-		for (const [fieldKey, field] of node.fields) {
-			const fieldSchema = schema.objectNodeFields.get(fieldKey);
-			if (fieldSchema === undefined) {
-				return SchemaValidationErrors.ObjectNode_FieldNotInSchema;
-			}
-			const fieldInSchemaResult = isFieldInSchema(
-				field,
-				fieldSchema,
-				nodeSchemaCollection,
-				schemaPolicy,
-			);
+		const fieldsActuallyInNode = new Set(node.fields.keys());
+		for (const [fieldKey, fieldSchema] of schema.objectNodeFields.entries()) {
```

Comment:
when iterating a map, you don't need to call entries. Maps are iterable:
```suggestion
		for (const [fieldKey, fieldSchema] of schema.objectNodeFields) {
```